mtz-ajax-interceptor-behavior.html

```diff
@@ -1,13 +1,136 @@
+   *  and calling the interceptors associated with those events.
+   *
+   * @polymerBehavior mtz.AjaxInterceptorBehavior
+   */
+  mtz.AjaxInterceptorBehavior = {
+    properties: {
+      /* The map of event to interceptor Set to listen against */
+      interceptors: {
+        type: Object,
+        value() {
+          return {};
+        }
+      }
+    },
+    attached() {
+      this.addEventListener('register-interceptor', this.__registerInterceptor);
+    },
+    detached() {
+      this.removeEventListener('register-interceptor', this.__registerInterceptor);
+      // Remove all active event listeners
+      Object.keys(this.interceptors).forEach((key) => {
+        this.__unregisterEventListener(key);
+      });
+    },
+    ready() {
+      // Ensures that when these functions are used they always get the proper this reference.
+      this.runInterceptors = this.runInterceptors.bind(this);
+      this.__registerInterceptor = this.__registerInterceptor.bind(this);
+      this.__unregisterInterceptor = this.__unregisterInterceptor.bind(this);
+    },
+    /**
+     * Runs all interceptors associated with the event, passing along the event to the interceptors chained together.
+     *
+     * @param {String} eventName - the event name that triggered the interceptor
+     * @param {Event} event - the event that triggered the interceptor logic
+     */
+    runInterceptors(eventName, event) {
+      const interceptors = this.interceptors[eventName];
+      if (!event || !interceptors) return;
 
-<script>
-  window.mtz = mtz || {};
+      // Override stopImmediatePropagation so we can break out if needed.
+      const stopImmediatePropagationFnc = event.stopImmediatePropagation;
+      let stopPropagation = false;
+      event.stopImmediatePropagation = () => {
+        stopPropagation = true;
+      };
+
+      for (const interceptor of Array.from(interceptors).sort((a, b) => b.priority - a.priority)) {
+        if (stopPropagation)
+          break;
+        interceptor.intercept(event, eventName);
+      }
+
+      // Restore stopImmediatePropagation incase it's needed.
+      event.stopImmediatePropagation = stopImmediatePropagationFnc;
+
+      // If propagation was stopped, ensure that the original logic is also called.
+      if (stopPropagation)
+        event.stopImmediatePropagation();
+    },
+    /**
+     * Registers an event listener for the provided event name against the runInterceptors method.
+     * @private
+     *
+     * @param {String} eventName
+     */
+    __registerEventListener(eventName) {
+      this.addEventListener(eventName, this.runInterceptors);
+    },
+    /**
+     * Adds the interceptor stored in event.detail to each event referenced in events, if a new event group is added
+     *  then __registerEventListener is called.
+     * @private
+     *
+     * @param {CustomEvent} event
+     */
+    __registerInterceptor(event) {
+      const events = event.detail.events;
```

Comment:
Is there a need for a check on event.detail.events being defined? This may output an error message in the dev tools console otherwise.